src/bucket.ts

```diff
@@ -988,6 +1003,136 @@ class Bucket extends ServiceObject {
     this.getFilesStream = paginator.streamify('getFiles');
   }
 
+  /**
+   * Set the object lifecycle rules for objects in this bucket.
+   *
+   * By default, the object lifecycle rule you provide to this method will be
+   * appended to any existing ones. To replace all existing rules, supply the
+   * `options` argument, setting `append` to `false`.
+   *
+   * @see [Object Lifecycle Management]{@link https://cloud.google.com/storage/docs/lifecycle}
+   * @see [Buckets: patch API Documentation]{@link https://cloud.google.com/storage/docs/json_api/v1/buckets/patch}
+   *
+   * @param {LifecycleRule} rule The new lifecycle rule to be added to objects
+   *     in this bucket.
+   * @param {AddLifecycleRuleOptions} [options] Configuration object.
+   * @param {SetBucketMetadataCallback} [callback] Callback function.
+   * @returns {Promise<SetBucketMetadataResponse>}
+   *
+   * @example
+   * const {Storage} = require('@google-cloud/storage');
+   * const storage = new Storage();
+   * const bucket = storage.bucket('albums');
+   *
+   * //-
+   * // Automatically have an object deleted from this bucket once it is 3 years
+   * // of age.
+   * //-
+   * bucket.addLifecycleRule({
+   *   action: 'delete',
+   *   condition: {
+   *     age: 365 * 3 // Specified in days.
+   *   }
+   * }, function(err, apiResponse) {});
+   *
+   * //-
+   * // For objects created before 2018, "downgrade" the storage class.
+   * //-
+   * bucket.addLifecycleRule({
+   *   action: 'setStorageClass',
+   *   condition: {
+   *     createdBefore: new Date('2018')
+   *   },
+   *   storageClass: 'COLDLINE'
+   * }, function(err, apiResponse) {});
+   *
+   * //-
+   * // Delete objects created before 2016 which have the Coldline storage
+   * // class.
+   * //-
+   * bucket.addLifecycleRule({
+   *   action: 'delete',
+   *   condition: {
+   *     matchesStorageClass: 'COLDLINE',
+   *     createdBefore: new Date('2016')
+   *   }
+   * }, function(err, apiResponse) {});
+   */
+  addLifecycleRule(rule: LifecycleRule): Promise<SetBucketMetadataResponse>;
+  addLifecycleRule(
```

Comment:
This overload is the same as the one at the bottom.  Pretty sure you can just get rid of this line (unless I am missing something)

Reply:
Here're all of 'em:

```js
addLifecycleRule(
  rule: LifecycleRule
): Promise<SetBucketMetadataResponse>;

addLifecycleRule(
  rule: LifecycleRule,
  optionsOrCallback?: AddLifecycleRuleOptions|SetBucketMetadataCallback
): Promise<SetBucketMetadataResponse>|void;

addLifecycleRule(
  rule: LifecycleRule,
  options: AddLifecycleRuleOptions,
  callback: SetBucketMetadataCallback
): void;

addLifecycleRule(
  rule: LifecycleRule,
  optionsOrCallback?: AddLifecycleRuleOptions|SetBucketMetadataCallback,
  callback?
): Promise<SetBucketMetadataResponse>|void;
```

Is there a better way?

Reply:
Yup! You should be able to delete the second one.

Reply:
If I do that, this fails:

```js
bucket.addLifecycleRule(
    {
      action: 'delete',
      condition: {
        age: 30,
        isLive: true,
      },
    },
    err => {
      assert.ifError(err);

      const rules = [].slice.call(bucket.metadata.lifecycle.rule);

      assert.deepStrictEqual(rules.pop(), {
        action: {
          type: 'Delete',
        },
        condition: {
          age: 30,
          isLive: true,
        },
      });

      done();
    });
```

![image](https://user-images.githubusercontent.com/723048/47305165-9bf50f80-d5f6-11e8-98c6-85237cae6841.png)

Reply:
ah, I see what's up.  You want this:

```js
addLifecycleRule(
  rule: LifecycleRule
  options?: AddLifecycleRuleOptions
): Promise<SetBucketMetadataResponse>;

addLifecycleRule(
  rule: LifecycleRule,
  options: AddLifecycleRuleOptions,
  callback: SetBucketMetadataCallback
): void;

addLifecycleRule(
  rule: LifecycleRule,
  callback: SetBucketMetadataCallback
): void;

addLifecycleRule(
  rule: LifecycleRule,
  optionsOrCallback?: AddLifecycleRuleOptions|SetBucketMetadataCallback,
  callback?
): Promise<SetBucketMetadataResponse>|void;
```